src/core/control/Control.cpp

```diff
@@ -1045,6 +1048,37 @@ void Control::clearSelectionEndText() {
     }
 }
 
+void Control::selectAllOnPage() {
+    auto pageNr = getCurrentPageNo();
+    if (pageNr == npos) {
+        return;
+    }
+
+    XojPageView* view = win->getXournal()->getViewFor(pageNr);
+    if (view == nullptr) {
+        return;
+    }
+
+    PageRef page = this->doc->getPage(pageNr);
+    Layer* layer = page->getSelectedLayer();
+
+    win->getXournal()->clearSelection();
+
+    std::vector<Element*> elements;
+    for (Element* e: layer->getElements()) { elements.push_back(e); }
```

Comment:
I tried it this way and for some reason it creates issues with selecting more than 2 things at once. This is the error I'm getting:
`** (xournalpp:31877): WARNING **: 14:27:31.405: Could not remove element from layer, it's not on the layer!`
and sometimes:
`** (xournalpp:31877): WARNING **: 14:38:16.294: Layer::insertElement() try to add an element twice!`
Then some items aren't selected or don't move properly after selection.

Assuming this is a bug, I think it's caused by how elements are ordered in the layer's element vector. This is also why I have to reverse the sort order (or at least re-instate it) in https://github.com/xournalpp/xournalpp/pull/4024#discussion_r872554054, because I'm reversing it here to get it to work. If someone fixed this we could make this change.

Reply:
Ah right, I see. The issue is here

https://github.com/xournalpp/xournalpp/blob/7da9a156b5e9bf9ee5167559966f7cde9c497321/src/core/control/tools/EditSelection.cpp#L88-L100

The elements get removed from the layer (so to our referenced vector) as they are added to the selection content. This is the reason for your messed up objects order too: in your situation, every call to `Layer::indexOf` return 0, a we already popped out the previous elements from the layer...

This is pretty inefficient in many respects:

- The copy aforementioned
- Successive calls to `Layer::removeElement` mean a complexity in O(n²) instead of O(n)
- Successive calls to `Layer::indexOf` mean a complexity in O(n²) instead of O(n) (we already know the index, since we are iterating on the layer content itself)

I think the most efficient solution would be:

1. To add a construct an empty `EditSelection` (using the first constructor) and do something like
```
Layer::Elementindex n = 0;
for (auto* e:layer->getElements()) {
    selection->addElement(e, n++);
}
layer->clear(false);
```
2. To add a `Layer::clear(bool free)`(used above) removing every element of a given layer (and optionally deleting them).

Alternatively, 1. could be a new constructor for the class `EditSelection` taking a `Layer*` as a parameter (among others).

Reply:
Thanks for the suggestions, everything seems to be working now.